### PR TITLE
Kafka Connect Avro Support Part 1: AvroSchemaManager

### PR DESCRIPTION
**Pull request by liuzix**
### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

### Release note

This is the first PR that will lead to support for outputting Avro-serialized data to Kafka. See more in #660. After research and deliberation, we gave up on syncing the Registry Schema ID through Etcd.

### CODE REVIEW

File: cdc/sink/schema_registry.go [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+	}
+
+	log.Info("Avro schema lookup cache miss",
+		zap.String("key", key),
+		zap.Int64("tiSchemaID", tiSchemaID))
+
+	uri := m.registryURL + "/subjects/" + url.QueryEscape(tableNameToSchemaSubject(tableName)) + "/versions/latest"
+	log.Debug("Querying for latest schema", zap.String("uri", uri))
+
+	req, err := http.NewRequest("GET", uri, nil)
```
Comment (overvenus) [inline cdc/sink/schema_registry.go:168]: Please leave a TODO about TLS support.

File: cdc/sink/schema_registry.go [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+	if err != nil {
+		return errors.Annotate(err, "Failed to read response from Registry")
+	}
+
+	if resp.StatusCode != 200 {
+		log.Warn("Failed to register schema to the Registry, HTTP error",
+			zap.Int("status", resp.StatusCode),
+			zap.String("uri", uri),
+			zap.ByteString("requestBody", payload),
+			zap.ByteString("responseBody", body))
```
Comment (overvenus) [inline cdc/sink/schema_registry.go:124]: Does it contain sensitive data? Eg. password or userdata.

File: cdc/sink/schema_registry.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+	var jsonResp registerResponse
+	err = json.Unmarshal(body, &jsonResp)
+
+	if err != nil {
+		return errors.Annotate(err, "Failed to parse result from Registry")
+	}
+
+	if jsonResp.ID == 0 {
+		return errors.New(fmt.Sprintf("Illegal schema ID returned from Registry %d", jsonResp.ID))
```
Comment (overvenus) [inline cdc/sink/schema_registry.go:136]:
```suggestion
		return errors.Errorf("Illegal schema ID returned from Registry %d", jsonResp.ID)
```

File: cdc/sink/schema_registry.go [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+	}
+
+	log.Info("Registered schema successfully",
+		zap.Int("id", jsonResp.ID),
+		zap.String("uri", uri),
+		zap.ByteString("body", body))
```
Comment (overvenus) [inline cdc/sink/schema_registry.go:142]: ByteString is not readable it produces something like "\u00FF". What about "zap.String"?

File: cdc/sink/schema_registry_test.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+       "name": "test",
+       "fields":
+         [
+           {
+             "type": "string",
+             "name": "field1"
+           },
+           {
+             "type": [
+      			"null",
```
Comment (overvenus) [inline cdc/sink/schema_registry_test.go:77]:
```suggestion
      	  "null",
```

Please avoid mixing tab and space.

File: cdc/sink/schema_registry.go [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+	return &AvroSchemaManager{
+		registryURL: registryURL,
+		cache:       make(map[string]*schemaCacheEntry, 1),
+	}, nil
+}
+
+var regexRemoveSpaces = regexp.MustCompile(`\s`)
+
+// Register the latest schema for a table to the Registry, by passing in a Codec
+func (m *AvroSchemaManager) Register(tableName model.TableName, codec *goavro.Codec) error {
```
Comment (overvenus) [inline cdc/sink/schema_registry.go:92]: Is there any go package handles registering schema?

File: cdc/sink/schema_registry.go (hunk 2 above)
Comment (liuzix) [reply cdc/sink/schema_registry.go:124]: If the user configures a registry URL that itself contains sensitive data, e.g. http://username:password@example.com, then it is possible. Other data are strictly about the schema, so they should not contain user data.

File: cdc/sink/schema_registry.go (hunk 4 above)
Comment (liuzix) [reply cdc/sink/schema_registry.go:142]: ByteString is readable when it contains only printable characters, and per my understanding, a JSON body should not contain any unprintable characters. If it does, then something suspicious is definitely happening, so some log output is necessary.

File: cdc/sink/schema_registry.go [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+		return nil
+	}
+
+	log.Error("Other error when clearing Registry")
+	return err
+}
+
+func tableNameToSchemaSubject(tableName model.TableName) string {
+	// We should guarantee unique names for subjects
+	return url.QueryEscape(tableName.Schema) + "%" + url.QueryEscape(tableName.Table)
```
Comment (amyangfei) [inline cdc/sink/schema_registry.go:258]: `%` could be in schema name or table, then both schema `t%`, table `t` and schema `t`, table `%t` will generate the same subject `t%%t`.
Recommend to use 
`model.QuoteSchema(schema, table)` before query escape

File: cdc/sink/schema_registry_test.go [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+package sink
+
+import (
+	"encoding/json"
+	"github.com/jarcoal/httpmock"
+	"github.com/linkedin/goavro/v2"
+	"github.com/pingcap/check"
+	"github.com/pingcap/ticdc/cdc/model"
+	"io/ioutil"
+	"net/http"
```
Comment (amyangfei) [inline cdc/sink/schema_registry_test.go:23]: Please reorg the import library as
```
standard lib

third party lib
```

File: cdc/sink/schema_registry.go (hunk 7 above)
Comment (liuzix) [reply cdc/sink/schema_registry.go:258]: I use QueryEscape first which replaces a plain "%" with "%25", thus avoiding the conflict. I will change tomodel.QuoteSchema() anyways.

File: cdc/sink/codec/schema_registry.go [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+			continue
+		}
+
+		if resp.StatusCode >= 200 || resp.StatusCode < 300 {
+			break
+		}
+		log.Warn("HTTP server returned with error", zap.Int("status", resp.StatusCode))
+		_ = resp.Body.Close()
+
+		time.Sleep(expBackoff.NextBackOff())
```
Comment (amyangfei) [inline cdc/sink/codec/schema_registry.go:284]: The backoff mechanism doesn't provide a max retry time

Comment (amyangfei) [reply cdc/sink/codec/schema_registry.go:284]: Can use `retry.Run` in `github.com/pingcap/ticdc/pkg/retry` for simple

**Comment by amyangfei** [comment]
Please fix data race in the unit test

File: cdc/sink/codec/schema_registry_test.go [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+	"github.com/pingcap/ticdc/cdc/model"
+)
+
+type AvroSchemaRegistrySuite struct {
+}
+
+var _ = check.SerialSuites(&AvroSchemaRegistrySuite{})
+
+type mockRegistry struct {
+	subjects map[string]*mockRegistrySchema
```
Comment (amyangfei) [inline cdc/sink/codec/schema_registry_test.go:35]: What about adding a lock to `mockRegistry` and protect the read/write of `subjects` to avoid data race.

**Pull request merged**